File: python-package/ChemicalDice/smiles_to_embeddings.py
import requests
import numpy as np
import pandas as pd
from tqdm import tqdm
import os
import math
import hashlib
import base64
from rdkit import Chem
import sys
import requests
import numpy as np
import pandas as pd
from tqdm import tqdm
import os
import math
import tempfile



# --- Configuration (must match the server) ---
# URL = "https://chemicaldice.ahujalab.iiitd.edu.in/stream-features-from-csv"
URL = "http://chemicaldice.ahujalab.iiitd.edu.in:8001/stream-features-from-csv"
BATCH_SIZE = 32
NUM_FEATURES = 8192
DTYPE = np.float32




def collect_features_from_csv(filepath: str,key: str ,convert_to_canonical: bool = True):
    """
    Collect feature embeddings from a CSV of SMILES strings by sending the file to a CDI
    API service and streaming back binary batches of numeric features.
    This function:
    - Loads a CSV file and requires a column named 'SMILES'.
    - Validates each SMILES string 
    - Optionally canonicalizes SMILES using process_smiles() and writes a temporary CSV
        containing the (possibly canonicalized) SMILES prior to upload.
    - Posts the CSV as multipart/form-data to a CDI API service endpoint (module-level
        constant URL), passing the provided API key via the "X-API-Key" header.
        returns a pandas.DataFrame where the first column is
        the SMILES strings and the remaining columns are features named "CDI1", "CDI2", ...
    Args:
            filepath (str): Path to the input CSV file. The CSV must contain a column named 'SMILES'.
            key (str): API key value to send in the "X-API-Key" request header.
            convert_to_canonical (bool, optional): If True (default), canonicalize SMILES using
                    process_smiles() before upload. Canonicalization results are written to a temporary
                    CSV which is submitted to the server; the original file is left untouched.
    Returns:
            pandas.DataFrame or None:
                    - On success: a DataFrame with shape (N, M+1), where N is the number of SMILES rows
                        in the input CSV and M == NUM_FEATURES is the number of numeric features per row.
                        The first column is 'SMILES' and the remaining columns are labeled 'CDI1'..'CDIM'.
                    - On network/request failure or if no batches are received: None is returned (after
                        printing an error or warning message).
                    - On invalid input (see Raises) an exception is raised instead of returning None.
    """

    df_data = pd.read_csv(filepath)
    if 'SMILES' not in df_data.columns:
        raise ValueError("CSV must contain a 'SMILES' column.")

    df_data['is_valid'] = df_data['SMILES'].apply(is_valid_smiles)
    
    num_invalid = (~df_data['is_valid']).sum()
    if num_invalid > 0:
        print(df_data[~df_data['is_valid']])
        print(f"Found {num_invalid} invalid SMILES. See above for details.")
        # print("There are invalid SMILES in the input CSV. Please fix or remove them before proceeding.")
        # Ask whether to continue with only valid SMILES or exit
        df_data.to_csv(filepath, index=False)
        df_data = df_data[df_data['is_valid']].reset_index(drop=True)
        print(f"Proceeding with {len(df_data)} valid SMILES.")
    else:
        print("All SMILES are valid.")


    if convert_to_canonical:
        print("Converting SMILES to canonical form...")
        df_data['SMILES'] = df_data['SMILES'].apply(process_smiles)
        # Save canonicalized dataframe to a temporary CSV file and update filepath to point to it
        tmp = tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.csv', newline='')
        tmp.close()
        df_data.to_csv(tmp.name, index=False)
        filepath = tmp.name
        print(f"Saved canonical SMILES to temp file: {filepath}")
    else:
        # Save canonicalized dataframe to a temporary CSV file and update filepath to point to it
        tmp = tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.csv', newline='')
        tmp.close()
        df_data.to_csv(tmp.name, index=False)
        filepath = tmp.name
        print(f"Saved canonical SMILES to temp file: {filepath}")

    NUM_ROWS = df_data.shape[0]
    # Calculate the size of one complete batch in bytes
    batch_byte_size = BATCH_SIZE * NUM_FEATURES * np.dtype(DTYPE).itemsize

    # Calculate expected number of batches for the progress bar
    total_batches = math.ceil(NUM_ROWS / BATCH_SIZE)
    # Calculate the size of one complete batch in bytes
    batch_byte_size = BATCH_SIZE * NUM_FEATURES * np.dtype(DTYPE).itemsize
    
    # Calculate expected number of batches for the progress bar
    total_batches = math.ceil(NUM_ROWS / BATCH_SIZE)
    headers = {"X-API-Key": key}
    received_batches = []
    try:
        # Open the local CSV file to be sent in the request
        with open(filepath, 'rb') as csv_file:
            # The 'files' dict tells requests to send a multipart/form-data POST
            # The key 'file' must match the argument name in the FastAPI endpoint
            files = {'file': (os.path.basename(filepath), csv_file, 'text/csv')}
            
            with requests.post(URL, files=files,headers=headers, stream=True) as response:
                response.raise_for_status()
                print(f"Sent {filepath}. Receiving stream...")

                progress_bar = tqdm(total=total_batches, unit="batch")

                for chunk in response.iter_content(chunk_size=batch_byte_size):
                    if chunk:
                        batch = np.frombuffer(chunk, dtype=DTYPE).reshape(BATCH_SIZE, NUM_FEATURES)
                        received_batches.append(batch)
                        progress_bar.update(1)
                
                progress_bar.close()

    except requests.exceptions.RequestException as e:
        print(f"Error during request: {e}")
        return

    if not received_batches:
        print("No batches were received.")
        return
    # Remove temporary file if it was created in the system temp directory
    try:
        tmpdir = os.path.abspath(tempfile.gettempdir())
        file_abspath = os.path.abspath(filepath)
        if file_abspath.startswith(tmpdir + os.sep) or file_abspath == tmpdir:
            os.remove(file_abspath)
    except Exception as e:
        print(f"Warning: could not remove temporary file {filepath}: {e}")
    # Assemble the final array
    print("\nStream finished. Concatenating batches...")

    if num_invalid > 0:
        print("Invalid SMILES were skipped. Check your input file which is_valid column where False indicates invalid SMILES.")
    final_array_with_padding = np.vstack(received_batches)
    
    # Trim any padding added to the last batch
    final_array = final_array_with_padding[:NUM_ROWS]

    # Convert the NumPy array to a DataFrame and prepend the SMILES column
    feature_cols = [f'CDI{i+1}' for i in range(final_array.shape[1])]
    df_features = pd.DataFrame(final_array, columns=feature_cols)
    df_features.insert(0, 'SMILES', df_data['SMILES'].values)

    return df_features




def process_smiles(s):
    mol = Chem.MolFromSmiles(str(s))
    if mol is None:
        print(f"Invalid SMILES: {s}")
        return None
    return Chem.MolToSmiles(mol, canonical=True)


def is_valid_smiles(smiles: str) -> bool:
    mol = Chem.MolFromSmiles(smiles)
    return mol is not None

File: R-package/R/smiles_to_embeddings.R
# client.R

# Load required libraries
library(httr)
library(progress)
library(jsonlite)
library(data.table)
library(reticulate)

# --- Configuration (must match the server) ---
BATCH_SIZE <- 32
NUM_FEATURES <- 8192 # Assuming this is your feature size
# In R, we specify the size in bytes. float32 is 4 bytes.
FLOAT_SIZE_BYTES <- 4
DTYPE_R <- "numeric" # The 'what' argument for readBin


# Before you run the function, you might need to tell reticulate which
# Python environment to use (the one where RDKit is installed).
# Import the necessary RDKit Python module into an R object
# rdkit <- import("rdkit.Chem", convert = TRUE)

#' Validates and canonicalizes a SMILES string using RDKit
#'
#' This function uses the Python RDKit library via reticulate to parse
#' a SMILES string. If the string is valid, it returns the canonical
#' SMILES. If it is invalid, it returns NA.
#'
#' @param smiles_string A character string representing a molecule.
#' @return The canonical SMILES string or NA_character_ if invalid.

process_smiles <- function(smiles_string) {
  rdkit <- import("rdkit.Chem", convert = TRUE)
  # Use tryCatch to handle errors from RDKit (for invalid SMILES)
  tryCatch({
    # 1. Convert the SMILES string to an RDKit molecule object
    mol <- rdkit$MolFromSmiles(smiles_string)
    
    # This check is crucial because MolFromSmiles returns NULL for invalid SMILES
    if (is.null(mol)) {
      stop("Invalid SMILES string") # Force an error to be caught
    }
    
    # 2. Convert the molecule object back to a canonical SMILES string
    canonical_smiles <- rdkit$MolToSmiles(mol)
    
    return(canonical_smiles)
    
  }, error = function(e) {
    # 3. If any step above fails, catch the error and return NA
    # message(paste("Invalid SMILES:", smiles_string, "->", e$message))
    return(NA_character_)
  })
}

is_valid_smiles <- function(smiles_string) {
  rdkit <- import("rdkit.Chem", convert = TRUE)
  tryCatch({
    mol <- rdkit$MolFromSmiles(smiles_string)
    if (is.null(mol)) {
      return(FALSE)
    }
    return(TRUE)
  }, error = function(e) {
    return(FALSE)
  })
}

# --- Main Function ---
collect_features_from_csv <- function(filepath, key , convert_to_canonical = TRUE) {
  
  # --- 1. Pre-process and Validate the Input CSV ---
  library(data.table)
  rdkit <- import("rdkit.Chem", convert = TRUE)
  message("Reading and validating CSV...")

  URL <- "https://chemicaldice.ahujalab.iiitd.edu.in/stream-features-from-csv"
  
  # Use data.table's fread for fast reading
  df_data <- fread(filepath)
  
  if (!"SMILES" %in% names(df_data)) {
    stop("CSV must contain a 'SMILES' column.")
  }
  
  # Apply the processing function to the SMILES column
  # Note: `lapply` returns a list, so we unlist it back to a vector
  #   df_data[, SMILES := unlist(lapply(SMILES, process_smiles))]

  df_data$is_valid <- unlist(lapply(df_data$SMILES, is_valid_smiles))

  num_invalid <- sum(!df_data$is_valid)


  if (num_invalid > 0) {
    # Print invalid rows
    print(df_data[!df_data$is_valid, ])
    
    cat(sprintf("Found %d invalid SMILES. See above for details.\n", num_invalid))
    # cat("There are invalid SMILES in the input CSV. Please fix or remove them before proceeding.\n")
    
    # Save dataframe to CSV
    write.csv(df_data, filepath, row.names = FALSE)
    
    # Keep only valid SMILES and reset rownames (like reset_index in pandas)
    df_data <- df_data[df_data$is_valid, , drop = FALSE]
    rownames(df_data) <- NULL
    
    cat(sprintf("Proceeding with %d valid SMILES.\n", nrow(df_data)))
    
    } else {
    cat("All SMILES are valid.\n")
    }


  
  # Assume df_data is a data.frame with a column "SMILES"
  # Assume process_smiles() is already defined (e.g., via reticulate with RDKit)

  if (convert_to_canonical) {
    cat("Converting SMILES to canonical form...\n")
    
    # Apply process_smiles to each SMILES string
    df_data$SMILES <- unlist(lapply(df_data$SMILES, process_smiles))
    
    # Save canonicalized dataframe to a temporary CSV file
    filepath <- tempfile(fileext = ".csv")
    write.csv(df_data, filepath, row.names = FALSE)

    #cat(sprintf("Coverted SMILES to  Canonical SMILES \n"))
    # cat(sprintf("Saved canonical SMILES to temp file: %s\n", filepath))
    
  } else {
    # Save dataframe to a temporary CSV file without canonicalization
    filepath <- tempfile(fileext = ".csv")
    write.csv(df_data[,c("SMILES"),drop=FALSE], filepath, row.names = FALSE)
    
    cat(sprintf("Saved SMILES to temp file: %s\n", filepath))
  }

  
  # Overwrite the original file with the processed data
  fwrite(df_data, filepath)
  
  # --- 2. Prepare for Streaming Request ---
  received_batches <- list()
  NUM_ROWS <- nrow(df_data)
  batch_byte_size <- as.integer(BATCH_SIZE * NUM_FEATURES * FLOAT_SIZE_BYTES)
  total_batches <- ceiling(NUM_ROWS / BATCH_SIZE)

  headers <- c()
  if (!is.null(key)) {
    headers <- add_headers(`X-API-Key` = key)
  }

  resp <- POST(
    URL,
    body = list(file = upload_file(filepath, type = "text/csv")),
    encode = "multipart",
    headers
  )

  if (http_error(resp)) {
    stop(sprintf("HTTP request failed: %s", status_code(resp)))
  }

  message(sprintf("Sent %s. Receiving stream...", filepath))
  raw_content <- content(resp, "raw")
  if (length(raw_content) == 0L) {
    message("No data received.")
    return(NULL)
  }

  buffer <- raw()
  pb <- progress_bar$new(format = "  [:bar] :percent (:current/:total) batches", total = total_batches, clear = FALSE, width = 60)

  byte_index <- 1L
  while (byte_index <= length(raw_content)) {
    # Append next chunk to buffer
    buffer <- c(buffer, raw_content[byte_index:min(length(raw_content), byte_index + batch_byte_size - 1L)])
    byte_index <- byte_index + batch_byte_size
    
    # Process all complete batches in buffer
    while (length(buffer) >= batch_byte_size) {
      chunk_raw <- buffer[1:batch_byte_size]
      buffer <- buffer[-(1:batch_byte_size)]
      
      con <- rawConnection(chunk_raw, open = "rb")
      values <- readBin(con, what = "numeric", n = BATCH_SIZE * NUM_FEATURES, size = 4, endian = "little")
      close(con)
      
      mat <- matrix(values, nrow = BATCH_SIZE, ncol = NUM_FEATURES, byrow = TRUE)
      received_batches[[length(received_batches) + 1]] <- mat
      pb$tick()
    }
  }
  pb$terminate()

  # Handle leftover partial batch
  if (length(buffer) > 0) {
    con <- rawConnection(buffer, open = "rb")
    values <- readBin(con, what = "numeric", n = length(buffer) / FLOAT_SIZE_BYTES, size = 4, endian = "little")
    close(con)
    rows <- length(values) / NUM_FEATURES
    if (rows > 0) {
      mat <- matrix(values, nrow = rows, ncol = NUM_FEATURES, byrow = TRUE)
      received_batches[[length(received_batches) + 1]] <- mat
    }
  }

  # Concatenate batches and trim padding to NUM_ROWS
  all_rows <- do.call(rbind, received_batches)
  final_array <- all_rows[seq_len(NUM_ROWS), , drop = FALSE]

  feature_cols <- paste0("CDI", seq_len(ncol(final_array)))
  df_features <- as.data.frame(final_array)
  colnames(df_features) <- feature_cols
  df_features <- cbind(SMILES = df_data$SMILES, df_features)

  message("\nStream finished. Concatenating batches...")
  if (num_invalid > 0) {
    cat("Invalid SMILES were skipped. Check your input file where is_valid == FALSE.\n")
  }
  return(df_features)
}



